### PR TITLE
[breaking-change] Yield settings in configure

### PR DESCRIPTION
**Pull request by Sija**
Changes how `#configure` yields given block, from `with self yield` to `yield settings`.

Before:

```cr
Klass.configure do
  settings.option = "value"
end
```

Now:

```cr
Klass.configure do |settings|
  settings.option = "value"
end
```

### CODE REVIEW

**Comment by Sija** [comment]
@paulcsmith ping?

**Comment by paulcsmith** [comment]
Thanks for the ping @Sija! I'm not quite sure that I'd like to include this. Can you write a bit about why you added this? I was thinking of going this route originally but then decided not to so that all applications using Habitat are consistent. 

Is the main reason for this that you prefer `config` over `settings`? If so I would consider switching it to `config` since it's more common and a bit shorter. I'd have to think about it and see what the community thinks too

**Comment by Sija** [comment]
Sure! My reasoning revolved around two points, namely:
- Familiarity of the construct with yielded object - a common pattern found in popular gems (airbreak, carrierwave) and elsewhere: [#1](https://robots.thoughtbot.com/mygem-configure-block), [#2](http://brandonhilkert.com/blog/ruby-gem-configuration-patterns/), [#3](https://kirillplatonov.com/2014/11/08/configure_ruby_module/)
  ```ruby
  Klass.configure do |settings|
    settings.option = "value"
  end
  ```

- [Plethora](https://github.com/crystal-lang/crystal/search?q=with+self+yield&type=Issues&utf8=%E2%9C%93) of bugs & quirks related to `with self yield` construct (some of them: https://github.com/crystal-lang/crystal/issues/4436, https://github.com/crystal-lang/crystal/issues/5117, https://github.com/crystal-lang/crystal/issues/5225) making this construct IMHO quite brittle, especially when the "main-selling-point" of it, DSL, is not really used, giving user essentially one method call: `#settings` - which can be easily made on the object yielded to the given block

---

Minor ones:
- What if I'd like to use different variable name (like before-mentioned `config`) for yielded `Settings` object?

- What about block forwarding? How should I access `#settings` while keeping `self` intact (if needed)?

**Comment by paulcsmith** [comment]
This makes a lot of sense. I need to think about it a bit more, but I think we should move forward with this and remove the `with self` completely and just yield the config/settings block. I don't see much advantage to yielding self anymore. Thoughts?

**Comment by Sija** [comment]
@paulcsmith Sounds like a swell idea to me :)

I'd only think about BC. OTOH changing:

```cr
Klass.configure do
  settings.option = "value"
end
```

into:

```cr
Klass.configure do |settings|
  settings.option = "value"
end
```

doesn't seem so terrible braking change.

Would you like me to update the PR with `with self` removal? Are they any other changes you'd like to see?

**Comment by Sija** [comment]
@paulcsmith Removed `with self` from `yield` call in `Habitat.configure`. Is there anything else to do before merging?

**Comment by Sija** [comment]
@paulcsmith ping?

**Review by paulcsmith — APPROVED** [review]
This looks great! I'm holding off on merging though so that I can make sure I've got everything in order for migrating the guides and the generated templates in Lucky CLI.

Thanks @Sija :D

**Comment by Sija** [comment]
@paulcsmith any timeframe on merging this?

**Comment by paulcsmith** [comment]
@sija probably when I do a new Lucky release, which will be in a few weeks (hopefully!)

Since this will require changing a bunch of templates in Lucky I'd rather do it in one go since a bunch of other stuff will change as well.

**Comment by Sija** [comment]
Cool, let me know if there's sth I could help :) Cheers!

**Comment by Sija** [comment]
@paulcsmith 🏓

**Comment by paulcsmith** [comment]
@Sija Thanks for the ping :) Gonna be another month I'd say. I want to get out this next release first because it's already really big! https://github.com/luckyframework/upgrade-diffs/commit/1ee884eb025daeede627c3b6a05fe78fde5229b4

Feel free to ping me again though if you're worried about it getting lost! I don't mind at all. I think I'll remember though because I definitely want to get this in at some point

**Comment by Sija** [comment]
Totally understandable.

**Comment by Sija** [comment]
@paulcsmith friendly 🏓 again :)

**Review by paulcsmith — APPROVED** [review]
Looks good :) Thanks for the ping!

**Pull request merged**
